pkg/collector/corechecks/containers/containerd_events.go

```diff
@@ -0,0 +1,294 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2018 Datadog, Inc.
+
+// +build containerd
+
+package containers
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	ctrUtil "github.com/DataDog/datadog-agent/pkg/util/containerd"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/api/events"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/gogo/protobuf/proto"
+)
+
+//ContainerdEvent contains the timestamp to make sure we flush all events that happened between two checks
+type ContainerdEvent struct {
+	ID        string
+	Timestamp time.Time
+	Topic     string
+	Namespace string
+	Message   string
+	Extra     map[string]string
+}
+
+type Subscriber struct {
+	sync.Mutex
+	Name                string
+	Filters             []string
+	Events              []ContainerdEvent
+	Namespace           string
+	CollectionTimestamp int64
+	isRunning           bool
+}
+
+func CreateEventSubscriber(name string, ns string, f []string) *Subscriber {
+	return &Subscriber{
+		Name:                name,
+		Namespace:           ns,
+		CollectionTimestamp: time.Now().Unix(),
+		Filters:             f,
+	}
+}
+
+func (s *Subscriber) CheckEvents(ctrItf ctrUtil.ContainerdItf) {
+	ctx := context.Background()
+	ev := ctrItf.GetEvents()
+	log.Info("Starting routine to collect Containerd events ...")
+	ctxNamespace := namespaces.WithNamespace(ctx, s.Namespace)
+	go s.run(ev, ctxNamespace)
+}
+
+// Run should only be called once, at start time.
+func (s *Subscriber) run(ev containerd.EventService, ctx context.Context) error {
+	stream, errC := ev.Subscribe(ctx, s.Filters...)
+	s.Lock()
+	s.isRunning = true
+	s.Unlock()
+	for {
+		select {
+		case message := <-stream:
+			switch message.Topic {
```

Comment:
To increase both this portion's testability and the overall `run` method's lifecycle readability, I would be tempted to isolate the parsing / conversion logic in a `processMessage(ev *Envelope) (ContainerdEvent, err)` method, WDYT?